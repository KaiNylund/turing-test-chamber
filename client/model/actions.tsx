import { AnyAction, ThunkDispatch } from "@reduxjs/toolkit";
import { RoleType } from "../types";
import { sendGuessToServer, sendMessageToServer } from "../utils/ServerInteractionUtils";

export function guessPlayer(player: "p1" | "p2") {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch({type: "guess/isGuessing", isGuessing: false});
    dispatch({type: "waiting", forEvent: "guessOutcome", value: true});
    dispatch({type: "waiting", forEvent: "gameStats", value: true});
    sendGuessToServer(player);
  }
}

export function inputTextChange(e: { target: { value: string; }; }) {
  let text = e.target.value;
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch({type: "input/textUpdate", message: text});
  }
}

export function submitMessage() {
  return async (dispatch: ThunkDispatch<{role: RoleType, inputText: string}, {}, AnyAction>,
                getState: () => {role: RoleType, inputText: string}) => {
    let curState = getState();
    let roleType = curState.role.type;
    let messageText = curState.inputText;
    // When we submit a new message, add it to the global state, and clear + disable the input
    // until we get a reply back.
    dispatch({type: "message/add", message: messageText, player: roleType});
    dispatch({type: "input/textUpdate", message: ""});
    dispatch({type: "waiting", forEvent: "playerMessage", value: true});
    sendMessageToServer(messageText, roleType);
  }
}

export function startGuessing() {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch({type: "guess/isGuessing", isGuessing: true});
  }
}