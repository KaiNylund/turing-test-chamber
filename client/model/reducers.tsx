import moment from 'moment';
import { combineReducers } from '@reduxjs/toolkit'
import { TextAction, PlayerAction, MessageAction, IsGuessingAction,
         GuessOutcomeAction } from "../types";

// safe version of structuredClone for node versions < 17
const structuredClone = (obj: any) => JSON.parse(JSON.stringify(obj))

export const defaultState = {
  messages: {
    p1: [],
    p2: [],
    guesser: []
  },
  // can be waiting for multiple things at once, so can't store waiting state as just a string
  // This way we can also subscribe to specific waiting conditions instead of checking whenever
  // a single waiting state changes.
  waitingFor: {
    serverCon: true,
    databaseCon: false,
    secondPlayer: false,
    role: false,
    playerMessage: false,
    gpt3Message: false,
    guessOutcome: false,
    gameStats: false
  },
  role: {
    type: "loading",
    assignedTime: ""
  },
  isGuessing: false,
  guessOutcome: "undecided",
  inputText: "",
  gameStats: null
};

function waitingReducer(state = defaultState.waitingFor,
                        action: {type: string, forEvent: "serverCon" | "databaseCon" | "role" |
                                                         "playerMessage" | "guessOutcome" |
                                                         "gameStats" | "gpt3Message",
                                 value: boolean}) {
  switch (action.type) {
    case "waiting":
      if (action.forEvent in state) {
        let stateCopy = structuredClone(state);
        stateCopy[action.forEvent] = action.value;
        return stateCopy;
      } else {
        return state;
      }
    case "waiting/clear":
      return {
        serverCon: false,
        databaseCon: false,
        secondPlayer: false,
        role: false,
        playerMessage: false,
        gpt3Message: false,
        guessOutcome: false,
        gameStats: false
      }
    default:
      return state;
  }
}

function guessOutcomeReducer(state = defaultState.guessOutcome, action: GuessOutcomeAction) {
  switch (action.type) {
    case "guess/outcome":
      return action.guessOutcome;
    default:
      return state
  }
}

function isGuessingReducer(state = defaultState.isGuessing, action: IsGuessingAction) {
  switch (action.type) {
    case "guess/isGuessing":
      return action.isGuessing;
    default:
      return state
  }
}

function inputTextReducer(state = defaultState.inputText, action: TextAction) {
  switch (action.type) {
    case "input/textUpdate":
      return action.message;
    default:
      return state
  }
}

function roleReducer(state = defaultState.role, action: PlayerAction) {
  switch (action.type) {
    case "role/assign":
      return {type: action.player, assignedTime: moment().utc().format()};
    default:
      return state;
  }
}

function messageReducer(state = defaultState.messages, action: MessageAction) {
  switch (action.type) {
    case "message/add":
      let stateCopy = structuredClone(state);
      if (action.player !== "loading") {
        let playerMessages: string[] = stateCopy[action.player];
        playerMessages.push(action.message);
      }
      return stateCopy;
    default:
      return state
  }
}

function gameStatsReducer(state = defaultState.gameStats, action: {type: string, gameStats: object}) {
  switch (action.type) {
    case "gameStats/update":
      return action.gameStats;
    default:
      return state;
  }
}

const appReducer = combineReducers({
  messages: messageReducer,
  role: roleReducer,
  inputText: inputTextReducer,
  waitingFor: waitingReducer,
  isGuessing: isGuessingReducer,
  guessOutcome: guessOutcomeReducer,
  gameStats: gameStatsReducer
});

export function rootReducer(state: any, action: any) {
  if (action.type === "reset") {
    return defaultState;
  }
  return appReducer(state, action);
}