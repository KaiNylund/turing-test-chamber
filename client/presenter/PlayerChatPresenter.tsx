import { Component, useState } from 'react';
import store from '../model/store';
import PlayerChatView from '../view/PlayerChatView';
import PlayerTitleView from '../view/PlayerTitleView';
import { DefaultStateType } from '../types';
import { connect } from 'react-redux';
import React from 'react';
import { guessPlayer } from '../model/actions';
import { animated, useSpring } from 'react-spring';

import Debug from "debug";
const debug = Debug("playerChatPresenter");

const mapStateToProps = (state: DefaultStateType) => {
  return {
    messages: state.messages,
    waitingForPlayerMessage: state.waitingFor.playerMessage,
    waitingForGpt3Message: state.waitingFor.gpt3Message
  }
}

const mapDispatchToProps = {
  guessPlayer
}

function PlayerChatPresenter(props: {player: "p1" | "p2",
                                     title: string,
                                     waitingForPlayerMessage: boolean,
                                     waitingForGpt3Message: boolean,
                                     messages: {p1: string[],
                                                p2: string[],
                                                guesser: string[]},
                                     guessPlayer: (player: "p1" | "p2") => void}) {
  const [title, setTitle] = useState(props.title);

  function onMouseOver() {
    if (store.getState().isGuessing) {
      setTitle(props.title + " is the machine");
    }
  }

  function onMouseLeave() {
    if (store.getState().isGuessing) {
      setTitle(props.title);
    }
  }

  function onClick() {
    let curState = store.getState();
    if (curState.isGuessing && curState.guessOutcome === "undecided") {
      debug("sending guess to server: " + props.player);
      props.guessPlayer(props.player);
      setTitle(props.title);
    }
  }

  let waitingForMessage: "none" | "message" | "reply" = "none";
  let role = store.getState().role.type;
  if (props.waitingForPlayerMessage && role === "guesser") {
    waitingForMessage = "reply";
  } else if (props.waitingForPlayerMessage && role !== "guesser") {
    waitingForMessage = "message";
  // If we're waiting for the gpt3 reply, and we're the tricker, and this is the gpt3 window
  // then add a waiting bubble on the reply side
  } else if (props.waitingForGpt3Message &&
             (role === "p1" || role === "p1") &&
             (role !== props.player)) {
    waitingForMessage = "reply";
  }

  const styleProps = useSpring({
    opacity: 1,
    transform: "translateY(0px)",
    from: { opacity: 0, transform: "translateY(-20px)" }
  });

  return (
    <animated.div className={"chat-window"} style={styleProps}>
      <PlayerTitleView title={title}/>
      <PlayerChatView messages={props.messages.guesser}
                      replies={props.messages[props.player]}
                      player={props.player}
                      waitingFor={waitingForMessage}
                      onMouseOver={onMouseOver}
                      onMouseLeave={onMouseLeave}
                      onClick={onClick}/>
    </animated.div>
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(PlayerChatPresenter);