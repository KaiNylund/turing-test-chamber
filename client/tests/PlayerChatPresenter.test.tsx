import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../model/store';
import PlayerChatPresenter from '../presenter/PlayerChatPresenter';
import { startGuessing } from '../model/actions';

const component = renderer.create(
  <Provider store={store}>
    <PlayerChatPresenter title={'p1'} player={'p1'}/>
  </Provider>
);

// Set the isGuessing flag to true so the playerChat's mouseOver and onClick functions will
// do something
store.dispatch({type: "guess/isGuessing", isGuessing: true});
let tree: any = component.toJSON();
let playerChatComp = tree.children.find((e: { type: string; }) => e.type === "ul");

// Test interactions with player chat when guessing
it('player chat hover and click effects work', async () => {
  expect(tree).toMatchSnapshot();

  // mouse over
  renderer.act(() => {
    playerChatComp.props.onMouseOver();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // mouse leave
  renderer.act(() => {
    playerChatComp.props.onMouseLeave();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // click
  renderer.act(() => {
    playerChatComp.props.onClick();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});