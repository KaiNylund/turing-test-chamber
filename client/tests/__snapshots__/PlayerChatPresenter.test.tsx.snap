// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player chat hover and click effects work 1`] = `
<div
  className="chat-window"
  style={
    {
      "opacity": 0,
      "transform": " translateY(-20px)",
    }
  }
>
  <h2>
    p1
  </h2>
  <ul
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`player chat hover and click effects work 2`] = `
<div
  className="chat-window"
  style={
    {
      "opacity": 0,
      "transform": " translateY(-20px)",
    }
  }
>
  <h2>
    p1 is the machine
  </h2>
  <ul
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`player chat hover and click effects work 3`] = `
<div
  className="chat-window"
  style={
    {
      "opacity": 0,
      "transform": " translateY(-20px)",
    }
  }
>
  <h2>
    p1
  </h2>
  <ul
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`player chat hover and click effects work 4`] = `
<div
  className="chat-window"
  style={
    {
      "opacity": 0,
      "transform": " translateY(-20px)",
    }
  }
>
  <h2>
    p1
  </h2>
  <ul
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;
